src/tests/tribol_enzyme_merged_jacobian.cpp

```diff
@@ -0,0 +1,656 @@
+    for ( int i{ 0 }; i < num_pairs; ++i ) {
+      auto elem_row_dofs = mfem::Array<int>();
+      fe_space_scalar.GetElementVDofs( lm_elems[i], elem_row_dofs );
+      auto elem_col_dofs = mfem::Array<int>();
+      fe_space.GetElementVDofs( nonmortar_elems[i], elem_col_dofs );
+      for ( int j{ 0 }; j < elem_row_dofs.Size(); ++j ) {
+        for ( int k{ 0 }; k < elem_col_dofs.Size(); ++k ) {
+          dgdn.Add( elem_row_dofs[j], elem_col_dofs[k], ( *elem_Js )[i]( j, k ) );
+        }
+      }
+    }
+    dgdn.Finalize();
+    // mfem::DenseMatrix dgdn_sparse;
+    // dgdn.ToDenseMatrix( dgdn_sparse );
+    // std::ofstream dgdn_file( "dgdn.mat" );
+    // dgdn_sparse.PrintMatlab( dgdn_file );
+    // dgdn_file.close();
+    auto dgdn_dndx = std::unique_ptr<mfem::SparseMatrix>( mfem::Mult( dgdn, dndx ) );
+    auto dgdx = std::unique_ptr<mfem::SparseMatrix>( mfem::Add( dgdx_nc, *dgdn_dndx ) );
+    mfem::DenseMatrix dgdx_enzyme;
+    dgdx->ToDenseMatrix( dgdx_enzyme );
+    dgdx.reset();
+
+    // dfdp (note: dndp = 0)
+    mfem::SparseMatrix dfdp( fe_space.GetVSize(), fe_space_scalar.GetVSize() );
+    // get mortar/lagrange contributions
+    elem_Js = &dfdx_data.getBlockJ()( static_cast<int>( BlockSpace::MORTAR ),
+                                      static_cast<int>( BlockSpace::LAGRANGE_MULTIPLIER ) );
+    for ( int i{ 0 }; i < num_pairs; ++i ) {
+      auto elem_row_dofs = mfem::Array<int>();
+      fe_space.GetElementVDofs( mortar_elems[i], elem_row_dofs );
+      auto elem_col_dofs = mfem::Array<int>();
+      fe_space_scalar.GetElementVDofs( lm_elems[i], elem_col_dofs );
+      for ( int j{ 0 }; j < elem_row_dofs.Size(); ++j ) {
+        for ( int k{ 0 }; k < elem_col_dofs.Size(); ++k ) {
+          dfdp.Add( elem_row_dofs[j], elem_col_dofs[k], ( *elem_Js )[i]( j, k ) );
+        }
+      }
+    }
+    // get nonmortar/lagrange contributions
+    elem_Js = &dfdx_data.getBlockJ()( static_cast<int>( BlockSpace::NONMORTAR ),
+                                      static_cast<int>( BlockSpace::LAGRANGE_MULTIPLIER ) );
+    for ( int i{ 0 }; i < num_pairs; ++i ) {
+      auto elem_row_dofs = mfem::Array<int>();
+      fe_space.GetElementVDofs( nonmortar_elems[i], elem_row_dofs );
+      auto elem_col_dofs = mfem::Array<int>();
+      fe_space_scalar.GetElementVDofs( lm_elems[i], elem_col_dofs );
+      for ( int j{ 0 }; j < elem_row_dofs.Size(); ++j ) {
+        for ( int k{ 0 }; k < elem_col_dofs.Size(); ++k ) {
+          dfdp.Add( elem_row_dofs[j], elem_col_dofs[k], ( *elem_Js )[i]( j, k ) );
+        }
+      }
+    }
+    dfdp.Finalize();
+    mfem::DenseMatrix dfdp_enzyme;
+    dfdp.ToDenseMatrix( dfdp_enzyme );
+    dfdp.Clear();
+
+    // save base gaps and forces
+    mfem::LinearForm gap_base( &fe_space_scalar );
+    gap_base = gap;
+    mfem::LinearForm force_base( &fe_space );
+    force_base = force;
+
+    // finite difference to obtain values to compare
```

Comment:
Can this utilize some of the code developed in FDCheck() in the tribol_enzyme_jacobian file?

Reply:
I worked on a finite difference class that we could use in these tests, but it didn't really work well, imo.  We can consider a better design in a follow on PR if needed.